### ISSUE TITLE
[Feature Request] Add element by element tensor reduction

### ISSUE THREAD

**Issue by vvvm23** [huggingface/accelerate#977]
Greetings,

## Motivation
I am currently refactoring my [VQ-VAE-2](https://github.com/vvvm23/vqvae-2/tree/refactor) repo to use :hugs: Accelerate (among other changes). I really want to use it in a data parallel context, but there is a missing feature in the library that is blocking this.

Although all-reduce ops exist in the library such as `accelerate.utils.reduce`, I want a method that preserves the dimensions of the tensors, simply performing the reduction element by element across processes. This is so I can calculate some statistics to update the vector quantisation codebook using EMA. The aforementioned function will instead collapse the tensor into a 0-dim tensor, which is not useful to me.

## Feature
To be explicit, suppose we have a tensor `x` with value `[1, 2, 3]` on process 1 and `[3,2,1]` on process 2, I want a function that can return the element-by-element all-reduce (assuming sum reduction) result `[4,4,4]`.

It would be nice to have other reductions available too.

The code could perhaps look like this:
```python
import accelerate
accelerator = accelerate.Accelerator()

# generate a per process tensor
x = torch.rand(4, 4)
x = accelerator.tensor_reduce(x)
# or
x = accelerator.reduce(x, keepdims=True)
# or, use the `accelerate.utils.reduce` version
print(x) # returns result with same shape as original tensor, but with updated values
```
I feel the latter fits better into the current API.

I could use `accelerator.gather` and then perform the sum after, however I think a dedicated `reduce` op would be more efficient.

I'm aware that this feature kinda goes beyond data parallelism into tensor parallelism. However, in my use case this just updates the codebook directly in the forward pass and does not take part in autograd at all. So despite this basically being a limited form of data parallelism, it is no more complicated that reduce ops we already do for metrics and the like. Hence, I think it could be in scope for `accelerate`.

Thanks!

**Comment by sgugger** [comment]
It would be valuable to have more reduction operations indeed. Would you like to open a PR to add this one? Relying on the existing gather to make sure it works on all kinds of distributed setups out of the box.

**Comment by vvvm23** [comment]
I can create a PR as an initial POC. Do you think the interface with `keepdims` is the best option?

However using gather (I think) is less efficient than using reduce ops as we will at some point store all `x` tensors to be reduced on all processes. Ideally we would do a real reduce using whatever chosen backend, however I wouldn't know how to do that.

**Comment by vvvm23** [comment]
Hi, any thoughts on the above? i.e. a separate function or try and integrate it with the existing `reduce` functions? Happy to make a PR but don't want to make something that doesn't fit with design philosphy of this library.

**Comment by pacman100** [comment]
Hello @vvvm23, I feel this is already supported. Below is the sample code and output:

1. Code: `test_reduce_elementwise.py`:
```
from accelerate import Accelerator
import torch


if __name__ == "__main__":
    accelerator = Accelerator()
    #print(accelerator.state)

    if accelerator.process_index == 0:
        tensor_to_reduce = torch.tensor([1.,2,3])
    else:
        tensor_to_reduce = torch.tensor([3.,4,0])
    tensor_to_reduce = tensor_to_reduce.to("cuda")
    #print(tensor_to_reduce)

    reduced_tensor = accelerator.reduce(tensor_to_reduce)
    print(reduced_tensor)
```

2. Output of command `accelerate launch --num_processes=2 --multi_gpu test_reduce_elementwise.py`:
```
[18:11:44] WARNING  The following values were not passed to `accelerate launch` and   launch.py:1074
                    had defaults used instead:                                                      
                            `--num_machines` was set to a value of `1`                              
                            `--mixed_precision` was set to a value of `'no'`                        
                            `--dynamo_backend` was set to a value of `'no'`                         
                    To avoid this warning pass in values for each of the problematic                
                    parameters or run `accelerate config`.                                          
tensor([4., 6., 3.], device='cuda:1')
tensor([4., 6., 3.], device='cuda:0')
```

Isn't this what you mentioned in the issue description?

**Comment by vvvm23** [comment]
> Hello @vvvm23, I feel this is already supported. Below is the sample code and output:
> 
>     1. Code: `test_reduce_elementwise.py`:
> 
> 
> ```
> from accelerate import Accelerator
> import torch
> 
> 
> if __name__ == "__main__":
>     accelerator = Accelerator()
>     #print(accelerator.state)
> 
>     if accelerator.process_index == 0:
>         tensor_to_reduce = torch.tensor([1.,2,3])
>     else:
>         tensor_to_reduce = torch.tensor([3.,4,0])
>     tensor_to_reduce = tensor_to_reduce.to("cuda")
>     #print(tensor_to_reduce)
> 
>     reduced_tensor = accelerator.reduce(tensor_to_reduce)
>     print(reduced_tensor)
> ```
> 
>     2. Output of command `accelerate launch --num_processes=2 --multi_gpu test_reduce_elementwise.py`:
> 
> 
> ```
> [18:11:44] WARNING  The following values were not passed to `accelerate launch` and   launch.py:1074
>                     had defaults used instead:                                                      
>                             `--num_machines` was set to a value of `1`                              
>                             `--mixed_precision` was set to a value of `'no'`                        
>                             `--dynamo_backend` was set to a value of `'no'`                         
>                     To avoid this warning pass in values for each of the problematic                
>                     parameters or run `accelerate config`.                                          
> tensor([4., 6., 3.], device='cuda:1')
> tensor([4., 6., 3.], device='cuda:0')
> ```
> 
> Isn't this what you mentioned in the issue description?

That is.. interesting. I am running on CPU as I don't currently have access to a GPU. If I remove the `.to("cuda")` part of your code I get a different result:
```
> accelerate launch --num_processes=2 --cpu test_reduce_elementwise.py
The following values were not passed to `accelerate launch` and had defaults used instead:
	`--num_machines` was set to a value of `1`
	`--mixed_precision` was set to a value of `'no'`
	`--dynamo_backend` was set to a value of `'no'`
To avoid this warning pass in values for each of the problematic parameters or run `accelerate config`.
tensor([1., 2., 3.])
tensor(6.)
```
I am using `accelerate==0.15.0`

**Comment by pacman100** [comment]
Hello @vvvm23, in the above example of CPU, could you uncomment `#print(accelerator.state)`. I feel `num_processes` is 1, wherein there is only one tensor and the reduce is simply summing that tensor which is expected behaviour

**Comment by vvvm23** [comment]
```
Num processes: 1
Process index: 0
Local process index: 0
Device: cpu
Mixed precision type: no
```
So, `num_processes` is indeed 1, despite me specifying 2.

> [...] the reduce is simply summing that tensor which is expected behaviour

I really feel like it shouldn't be the expected behaviour. If the point of `accelerate` is to separate model code and the like from parallel execution strategies, why does the shape of the output vary depending on the number of processes? It feels strange to me that the output of an operation in my model changes depending on the execution strategy, when the idea is to abtract away this complexity.

It should be worth saying, that when I originally made this issue I was testing on (a single) GPU, so this isn't a CPU only issue.

**Comment by pacman100** [comment]
> To be explicit, suppose we have a tensor x with value [1, 2, 3] on process 1 and [3,2,1] on process 2, I want a function that can return the element-by-element all-reduce (assuming sum reduction) result [4,4,4]

Hello, if it was single GPU, then there is no concept of multiple processes but that was the main point mentioned in issue description. I think I am confused wrt what the issue is. Could you please provide a minimal reproducible example for us to deep dive.

**Comment by muellerzr** [comment]
@vvvm23 to see just what our current reduce implementation is, I've dropped it below. You can see that on a single GPU (or on CPU), it calls `.sum()` or `.mean()` depending on the reduction specified, otherwise calls the appropriate reduction otherwise:

```python
        if state.distributed_type == DistributedType.TPU:
            xm.all_reduce("sum", cloned_tensor)
            return cloned_tensor
        elif state.distributed_type.value in CUDA_DISTRIBUTED_TYPES:
            torch.distributed.all_reduce(cloned_tensor, ReduceOp.SUM)
            return cloned_tensor
        else:
            if reduction == "sum":
                return cloned_tensor.sum()
            else:
                return cloned_tensor.mean()
```

I am also now a bit confused as to what exactly is the need :) https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/operations.py#L416-L446

**Comment by vvvm23** [comment]
Sorry yes, I think I confused myself too :,)

I think basically I was expecting zero code change when running the same code in a single or multi process context. So, I can develop my code on a single GPU and wrap any tensors that would need to be reduced in the future in a `accelerator.reduce(...)`.

So, when only one process is in use, I expected it to behave as an identity operation. When multiple are in use, I expected it to perform the cross-process reduction. In any case, I expected the output shape to remain the same on all processes, so not to break any future computation using that tensor. However, depending on the number of processes the shape can remain the same, or collapse to a 0-dim tensor.

I think the confusion arose because I found this behaviour when developing on one GPU, but as I don't have multiple GPUs I executed the reproducer code I posted on multiple CPUs - which as we just saw does not actually seem to use multiple processes, even when it is specified to do so as a launch parameter.

**Comment by vvvm23** [comment]
To put it into context of what I was trying to achieve in my project, take a look at [the commented out lines here](https://github.com/vvvm23/vqvae-2/blob/51d0080d14f1fc453362df3b943deec5419976c3/vqvae2/vq.py#L58)

What I see when I uncomment that line is `embedding_onehot_sum` become a 0-dim tensor when I want it to remain as a n-dim tensor. From what I understand from the discussion so far, if I were to run the same code on 2 processes, it would now return a n-dim tensor - the correct answer.

This feels strange to me, as now I need to write extra code to handle the single process case when I thought the idea for `accelerate` is to abstract away logic like this.
